### PR TITLE
Add test scenarios txpool - Closes #5046

### LINKED ISSUE #5046: Add more test scenarios to TxPool
### Expected behavior

All the test scenarios should be covered for the Transaction pool.

### Actual behavior

Some missing scenarios.

### Steps to reproduce

`npm t`

### Which version(s) does this affect? (Environment, OS, etc...)

`4.0.0`

### LINKED ISSUE #5047: DEFAULT_REPLACEMENT_FEE_DIFF should default to 10 in TransactionList
### Expected behavior
There should be only one default value for DEFAULT_MINIMUM_REPLACEMENT_FEE_DIFFERENCE in the TransactionPool.

### Actual behavior

A duplicate value DEFAULT_REPLACEMENT_FEE_DIFF defaults to 0 in the TransactionList

### Steps to reproduce

`npm t`

### Which version(s) does this affect? (Environment, OS, etc...)

`4.0.0`

### PR DESCRIPTION
**Pull request by ishantiw**
### What was the problem?

This PR resolves #5046 & resolves  #5047 

### How was it solved?

- [x] DEFAULT_MINIMUM_REPLACEMENT_FEE_DIFFERENCE in TransactionList is equal to `10`
- [x] `exceededTransactionsCount >= 0` to trigger eviction when the TxPool size is equal maxSize
- [x]  **Test Scenario:** when the same nonce transaction with higher fee and greater than minReplaceFeeDiff is added
- [x] **Test Scenario:** should evict the trx from the pool when txPool is full when all the trxs are processable
- [x] **Test Scenario:** should evict the trx from the pool when txPool is full when not all of the trxs are processable
- [x] Use multiple addresses and their transactions for test cases

### How was it tested?

Run `npm t` under `/elements/lisk-transaction-pool`

### CODE REVIEW

File: elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts [hunk 1]
```diff
@@ -210,7 +210,7 @@ describe('TransactionList class', () => {
 			});
 		});
 
-		describe('given list has no space', () => {
+		describe('given list of trxs with continuous nonce', () => {
```
Comment (shuse2) [inline elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts]: I think this setup was for `when the transaction list is full` cases

File: elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts [hunk 2]
```diff
@@ -378,7 +398,7 @@ describe('TransactionList class', () => {
 	});
 
 	describe('promote', () => {
-		describe('when promoting transaction does exist id', () => {
+		describe('when promoting transaction does not exist id', () => {
```
Comment (shuse2) [inline elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts]:
```suggestion
		describe('when promoting transaction id does not exist', () => {
```

File: elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts [hunk 3]
```diff
@@ -343,6 +343,133 @@ describe('TransactionPool class', () => {
...
+			tempApplyTransactionsStub.mockResolvedValue([
+				{ status: Status.OK, errors: [] },
+			]);
+			jest.spyOn(transactionPool, '_evictProcessable' as any);
+			const { status } = await transactionPool.add(highFeePriorityTx);
+			expect(transactionPool['_evictProcessable']).toHaveBeenCalledTimes(1);
+			expect(status).toEqual(Status.OK);
+		});
+
+		it('should evict the trx from the pool when txPool is full when not all of the trxs are processable', async() => {
```
Comment (shuse2) [inline elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts]:
```suggestion
		it('should evict unprocessable trx from the pool when txPool is full when not all of the trxs are processable', async() => {
```

File: elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts [hunk 4]
```diff
@@ -343,6 +343,133 @@ describe('TransactionPool class', () => {
 			expect(status).toEqual(Status.FAIL);
 		});
 
+		it('should evict the trx from the pool when txPool is full when all the trxs are processable', async() => {
```
Comment (shuse2) [inline elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts]:
```suggestion
		it('should evict lowest feePriotity among highest nonce trx from the pool when txPool is full when all the trxs are processable', async() => {
```

Comment (mitsuaki-u) [reply elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts]: Good suggestion maybe also change `when all the trxs are processable` to `and all the trxs are processable`

File: elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts [hunk 5]
```diff
@@ -310,7 +330,7 @@ describe('TransactionList class', () => {
 				});
 			});
 
-			describe('when new transaction is added with higher nonce than the existing highest nonce', () => {
+			describe('when new transaction is added with higher nonce than the existing highest nonce when transactionList.size >= maxSize', () => {
```
Comment (IkerAlus) [inline elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts]: As Shu pointed out before, we are already inside the case, so `when transactionList.size >= maxSize` is not necessary here

File: elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts [hunk 6]
```diff
@@ -327,7 +347,7 @@ describe('TransactionList class', () => {
 				});
 			});
 
-			describe('when new transaction is added with lower nonce than the existing highest nonce', () => {
+			describe('when new transaction is added with lower nonce than the existing highest nonce when transactionList.size >= maxSize', () => {
```
Comment (IkerAlus) [inline elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts]: As Shu pointed out before, we are already inside the case, so `when transactionList.size >= maxSize` is not necessary here

File: elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts [hunk 7]
```diff
@@ -541,7 +561,7 @@ describe('TransactionList class', () => {
 			});
 		});
 
-		describe('when there are only unprocessable transactions and all nonce are partially continuous', () => {
+		describe('when there are only unprocessable transactions and all nonce are in sequence', () => {
```
Comment (IkerAlus) [inline elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts]:
```suggestion
		describe('when there are only unprocessable transactions and all nonces are in sequence', () => {
```

Comment (IkerAlus) [reply elements/lisk-transaction-pool/test/unit/transaction_list.spec.ts]: the plural of nonce is also missing in the previous descriptions of `getPromotable` tests.

File: elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts [hunk 8]
```diff
@@ -343,6 +343,136 @@ describe('TransactionPool class', () => {
...
+			tempApplyTransactionsStub.mockResolvedValue([
+				{ status: Status.OK, errors: [] },
+			]);
+			jest.spyOn(transactionPool, '_evictProcessable' as any);
+			const { status } = await transactionPool.add(highFeePriorityTx);
+			expect(transactionPool['_evictProcessable']).toHaveBeenCalledTimes(1);
+			expect(status).toEqual(Status.OK);
+		});
+
+		it('should evict unprocessable trx from the pool when txPool is full and not all trxs are processable', async () => {
```
Comment (IkerAlus) [inline elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts]: shouldn't this be:
```
should evict the unprocessable trx with the lowest feePriority from the pool when txPool is full and not all trxs are processable
```
?

File: elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts [hunk 9]
```diff
@@ -343,6 +343,136 @@ describe('TransactionPool class', () => {
 			expect(status).toEqual(Status.FAIL);
 		});
 
+		it('should evict lowest feePriotity among highest nonce trx from the pool when txPool is full and all the trxs are processable', async () => {
```
Comment (IkerAlus) [inline elements/lisk-transaction-pool/test/unit/transaction_pool.spec.ts]:
```suggestion
		it('should evict lowest feePriority among highest nonce trx from the pool when txPool is full and all the trxs are processable', async () => {
```

**Comment by ishantiw** [comment]
@IkerAlus 
Thanks for your comments.
Regarding your below comment, I think we already discussed in the review meeting but I missed make that change. Its a duplicate test case, maybe the intention was to test it with a number or string in the id but in both the cases it is only a string. @shuse2 ?

> transaction_list.spec.ts: Lines 418: what is the difference between this test case and the one at line 401? Also, the description of it is wrong (refer to Shu's comment at line 401).

**Pull request merged**